File: scripts/js/fetch-pool-data.js
#!/usr/bin/env node

/**
 * Haal alle PoolStand en PoolStandRegel data op voor seizoen 2025-2026
 * Output: CSV en JSON met team-info, punten, positie, etc.
 */

import { PrismaClient } from "@oranje-wit/database";

const prisma = new PrismaClient();

async function main() {
  const seizoen = "2025-2026";

  console.log(`\n=== PoolStand data ophalen voor seizoen ${seizoen} ===\n`);

  // Query: alle PoolStands met gerelateerde regels
  const poolStanden = await prisma.poolStand.findMany({
    where: { seizoen },
    include: {
      regels: {
        orderBy: { positie: "asc" },
      },
    },
    orderBy: [{ periode: "asc" }, { niveau: "asc" }, { pool: "asc" }],
  });

  console.log(`Gevonden: ${poolStanden.length} pools\n`);

  if (poolStanden.length === 0) {
    console.log("Geen pools gevonden voor dit seizoen.");
    await prisma.$disconnect();
    process.exit(0);
  }

  // Aggregeer data per niveau
  const niveaus = {};
  const alleRegels = [];

  for (const poolStand of poolStanden) {
    const key = `${poolStand.periode} - ${poolStand.niveau || "onbekend"}`;
    if (!niveaus[key]) {
      niveaus[key] = {
        periode: poolStand.periode,
        niveau: poolStand.niveau,
        pools: [],
        teams: [],
        totaalTeams: 0,
        totaalOWTeams: 0,
      };
    }

    niveaus[key].pools.push(poolStand.pool);

    for (const regel of poolStand.regels) {
      niveaus[key].teams.push(regel.teamNaam);
      niveaus[key].totaalTeams++;
      if (regel.isOW) {
        niveaus[key].totaalOWTeams++;
      }

      alleRegels.push({
        periode: poolStand.periode,
        pool: poolStand.pool,
        niveau: poolStand.niveau || "onbekend",
        regio: poolStand.regio || "",
        positie: regel.positie,
        teamNaam: regel.teamNaam,
        isOW: regel.isOW ? "JA" : "NEE",
        gespeeld: regel.gespeeld,
        gewonnen: regel.gewonnen,
        gelijk: regel.gelijk,
        verloren: regel.verloren,
        punten: regel.punten,
        doelpuntenVoor: regel.doelpuntenVoor,
        doelpuntenTegen: regel.doelpuntenTegen,
        doelverschil: regel.doelpuntenVoor - regel.doelpuntenTegen,
      });
    }
  }

  // Samenvatting per niveau
  console.log("=== SAMENVATTING PER NIVEAU/PERIODE ===\n");
  for (const [key, data] of Object.entries(niveaus).sort()) {
    console.log(`${key}`);
    console.log(`  Pools: ${data.pools.join(", ")}`);
    console.log(`  Teams: ${data.totaalTeams} (waarvan ${data.totaalOWTeams} OW)\n`);
  }

  // Alle regels uitgebreid
  console.log("\n=== VOLLEDIGE TEAMDATA ===\n");
  console.log(
    "periode,pool,niveau,regio,positie,teamNaam,isOW,gespeeld,gewonnen,gelijk,verloren,punten,doelpuntenVoor,doelpuntenTegen,doelverschil"
  );

  for (const regel of alleRegels) {
    console.log(
      `${regel.periode},${regel.pool},${regel.niveau},${regel.regio},${regel.positie},${regel.teamNaam},${regel.isOW},${regel.gespeeld},${regel.gewonnen},${regel.gelijk},${regel.verloren},${regel.punten},${regel.doelpuntenVoor},${regel.doelpuntenTegen},${regel.doelverschil}`
    );
  }

  // JSON export
  console.log("\n\n=== JSON EXPORT ===\n");
  console.log(
    JSON.stringify(
      {
        seizoen,
        totaalPools: poolStanden.length,
        totaalTeams: alleRegels.length,
        owTeams: alleRegels.filter((r) => r.isOW === "JA").length,
        data: alleRegels,
      },
      null,
      2
    )
  );

  // Statistieken per niveau
  console.log("\n\n=== STATISTIEKEN PER NIVEAU ===\n");
  const niveauStats = {};
  for (const regel of alleRegels) {
    if (!niveauStats[regel.niveau]) {
      niveauStats[regel.niveau] = {
        teams: 0,
        owTeams: 0,
        gemiddeldePunten: [],
        gemiddeldeGespeeld: [],
      };
    }
    niveauStats[regel.niveau].teams++;
    if (regel.isOW === "JA") niveauStats[regel.niveau].owTeams++;
    niveauStats[regel.niveau].gemiddeldePunten.push(regel.punten);
    niveauStats[regel.niveau].gemiddeldeGespeeld.push(regel.gespeeld);
  }

  for (const [niveau, stats] of Object.entries(niveauStats).sort()) {
    const gemPunten = (stats.gemiddeldePunten.reduce((a, b) => a + b, 0) / stats.teams).toFixed(1);
    const gemGespeeld = (stats.gemiddeldeGespeeld.reduce((a, b) => a + b, 0) / stats.teams).toFixed(
      1
    );
    console.log(
      `${niveau}: ${stats.teams} teams (${stats.owTeams} OW) | gem. punten: ${gemPunten} | gem. gespeeld: ${gemGespeeld}`
    );
  }

  await prisma.$disconnect();
}

main().catch((error) => {
  console.error("Fout:", error);
  process.exit(1);
});

File: scripts/python/fetch-pool-data.py
#!/usr/bin/env python3
"""
Haal alle PoolStand en PoolStandRegel data op uit PostgreSQL
voor seizoen 2025-2026 en 2024-2025.
"""

import os
import json
import csv
import sys
from io import StringIO
from collections import defaultdict

try:
    import psycopg2
    from psycopg2.extras import DictCursor
except ImportError:
    print("⚠️  psycopg2 niet geïnstalleerd. Probeer: pip install psycopg2-binary")
    sys.exit(1)

# Database connection
DATABASE_URL = os.environ.get("DATABASE_URL")
if not DATABASE_URL:
    print("ERROR: DATABASE_URL niet ingesteld in .env")
    sys.exit(1)

def fetch_pool_data(seizoen):
    """Haal alle pool-data op voor een seizoen"""
    try:
        conn = psycopg2.connect(DATABASE_URL)
        cur = conn.cursor(cursor_factory=DictCursor)

        query = """
        SELECT
          ps.seizoen,
          ps.periode,
          ps.pool,
          ps.niveau,
          ps.regio,
          ps.stand_datum,
          psr.positie,
          psr.team_naam,
          psr.is_ow,
          psr.gs as gespeeld,
          psr.wn as gewonnen,
          psr.gl as gelijk,
          psr.vl as verloren,
          psr.pt as punten,
          psr.vr as doelpunten_voor,
          psr.tg as doelpunten_tegen
        FROM pool_standen ps
        LEFT JOIN pool_stand_regels psr ON ps.id = psr.pool_stand_id
        WHERE ps.seizoen = %s
        ORDER BY ps.periode, ps.niveau, ps.pool, psr.positie
        """

        cur.execute(query, (seizoen,))
        rows = cur.fetchall()
        cur.close()
        conn.close()

        return rows
    except Exception as e:
        print(f"ERROR: Kan database niet bereiken: {e}", file=sys.stderr)
        sys.exit(1)

def main():
    for seizoen in ["2025-2026", "2024-2025"]:
        print(f"\n{'='*70}")
        print(f"PoolStand data voor {seizoen}")
        print(f"{'='*70}\n")

        rows = fetch_pool_data(seizoen)

        if not rows:
            print(f"⚠️  Geen data gevonden voor {seizoen}")
            continue

        print(f"✓ {len(rows)} teamrecords opgehaald\n")

        # Aggregeer per niveau
        niveaus = defaultdict(lambda: {
            'teams': 0,
            'ow_teams': 0,
            'pools': set(),
            'min_punten': float('inf'),
            'max_punten': float('-inf'),
            'punten': [],
            'gespeeld': []
        })

        # CSV output
        print("PERIODE,POOL,NIVEAU,REGIO,STAND_DATUM,POSITIE,TEAM_NAAM,IS_OW,GESPEELD,GEWONNEN,GELIJK,VERLOREN,PUNTEN,DOELPUNTEN_VOOR,DOELPUNTEN_TEGEN,DOELVERSCHIL")

        for row in rows:
            if row['team_naam']:
                # CSV line
                doelverschil = (row['doelpunten_voor'] or 0) - (row['doelpunten_tegen'] or 0)
                print(f"{row['periode']},{row['pool']},{row['niveau'] or ''},{row['regio'] or ''},{row['stand_datum'] or ''},"\
                      f"{row['positie']},{row['team_naam']},{'JA' if row['is_ow'] else 'NEE'},"\
                      f"{row['gespeeld']},{row['gewonnen']},{row['gelijk']},{row['verloren']},"\
                      f"{row['punten']},{row['doelpunten_voor']},{row['doelpunten_tegen']},{doelverschil}")

                # Aggregatie
                niveau = row['niveau'] or 'onbekend'
                niveaus[niveau]['teams'] += 1
                if row['is_ow']:
                    niveaus[niveau]['ow_teams'] += 1
                niveaus[niveau]['pools'].add(row['pool'])
                if row['punten']:
                    niveaus[niveau]['min_punten'] = min(niveaus[niveau]['min_punten'], row['punten'])
                    niveaus[niveau]['max_punten'] = max(niveaus[niveau]['max_punten'], row['punten'])
                    niveaus[niveau]['punten'].append(row['punten'])
                if row['gespeeld']:
                    niveaus[niveau]['gespeeld'].append(row['gespeeld'])

        # Statistieken
        print(f"\n\n=== STATISTIEKEN PER NIVEAU ===\n")
        for niveau in sorted(niveaus.keys()):
            stats = niveaus[niveau]
            gem_punten = sum(stats['punten']) / len(stats['punten']) if stats['punten'] else 0
            gem_gespeeld = sum(stats['gespeeld']) / len(stats['gespeeld']) if stats['gespeeld'] else 0
            pools_str = ', '.join(sorted(stats['pools']))

            print(f"{niveau}")
            print(f"  Teams: {stats['teams']} (waarvan {stats['ow_teams']} OW)")
            print(f"  Pools: {pools_str}")
            print(f"  Punten: min={stats['min_punten']}, max={stats['max_punten']}, gem={gem_punten:.1f}")
            print(f"  Gespeeld: gem={gem_gespeeld:.1f}\n")

if __name__ == "__main__":
    main()
